torch/testing/_internal/distributed/rpc/jit/rpc_test.py

```diff
@@ -58,10 +69,352 @@ def forward(self):
+            ):
+                args = (
+                    torch.tensor([1, 1]),
+                    torch.tensor([2, 2]),
+                    torch.tensor([3, 3]),
+                    torch.tensor([4, 4]),
+                    torch.tensor([5, 5]),
+                )
+                kwargs = {}
+                fut = rpc.api.rpc_async(
+                    dst_worker_name, two_args_two_kwargs, args, kwargs
+                )
+                ret = fut.wait()
+                return ret
+
+    @dist_init
+    def test_unexepected_kwarg_is_specified(self):
+        if self.rank != 0:
+            return
+
+        dst_worker_name = "worker{}".format((self.rank + 1) % self.world_size)
+
+        # Notice, kwargs matching happens during execution.
+        @torch.jit.script
+        def rpc_async_call_remote_torchscript_in_torchscript_with_unexpected_kwarg(
+            dst_worker_name: str,  # noqa: E999
+        ):
+            args = (torch.tensor([1, 1]), torch.tensor([2, 2]))
+            kwargs = {"third_kwarg": torch.tensor([1, 1])}
+            fut = rpc.api.rpc_async(dst_worker_name, two_args_two_kwargs, args, kwargs)
+            ret = fut.wait()
+            return ret
+
+        with self.assertRaisesRegex(
+            RuntimeError, "Unknown keyword argument 'third_kwarg'"
+        ):
+            ret = rpc_async_call_remote_torchscript_in_torchscript_with_unexpected_kwarg(
+                dst_worker_name
+            )
+            self.assertEqual(ret, 0)
+
+    @dist_init
+    def test_call_python_function_remotely_from_script_not_supported(self):
+        if self.rank != 0:
+            return
+
+        dst_worker_name = "worker{}".format((self.rank + 1) % self.world_size)
+
+        @torch.jit.script
+        def rpc_async_call_remote_py_function_in_torchscript(dst_worker_name: str):
+            args = ()
+            kwargs = {}
+            fut = rpc.api.rpc_async(dst_worker_name, python_function, args, kwargs)
+            ret = fut.wait()
+            return ret
+
+        with self.assertRaisesRegex(
+            RuntimeError, "attempted to get undefined function"
+        ):
+            ret = rpc_async_call_remote_py_function_in_torchscript(dst_worker_name)
+            self.assertEqual(ret, 0)
+
+    @dist_init
+    def test_call_script_function_that_raises_remotely_from_script(self):
+        if self.rank != 0:
+            return
+
+        dst_worker_name = "worker{}".format((self.rank + 1) % self.world_size)
+
+        # Notice, TorchScript always translates(emits) Python `raise` statement,
+        # as the exception message string, "Exception",
+        # no matter what exception type and excetpion message are in the statement,
+        @torch.jit.script
+        def rpc_async_call_remote_raising_torchscript_in_torchscript(
+            dst_worker_name: str
+        ):
+            args = ()
+            kwargs = {}
+            fut = rpc.api.rpc_async(dst_worker_name, raise_script, args, kwargs)
```

Comment:
I'm a little confused with the API changes here. Have a bunch of questions about the API and it would be good mention this in the PR summary:

1. What is the difference between this rpc_async method and the one in the Python file rpc/api.py?
2. Does the user need to specify `rpc.api`? If so can we just make it `rpc.rpc_sync`?
3. Are we getting rid of `_rpc_async_torchscript`? If so, why is it still there in rpc/api.py?

Reply:
@pritamdamania87 

> What is the difference between this rpc_async method and the one in the Python file rpc/api.py?

Since in `rpc/__init__.py`, we wrote ` from .api import rpc_async`, so `rpc.rpc_async` and `rpc.api.rpc_async` point to the same object. 


> Does the user need to specify rpc.api? If so can we just make it rpc.rpc_async?

Yeah, it's confusing and setting up a bad examples for users, let me update it. 

> Are we getting rid of _rpc_async_torchscript? If so, why is it still there in rpc/api.py?

Answered in https://github.com/pytorch/pytorch/pull/33329#discussion_r381557338.

I can remove `_rpc_sync_torchscript` and `_rpc_async_torchscript` together in this PR.